### PR TITLE
Ion List and Slide Box

### PR DESCRIPTION
**Pull request by JFickel**

### CODE REVIEW

File: addon/components/ion-button.js [hunk 1]
```diff
@@ -13,10 +13,13 @@ export default Ember.Component.extend({
     'clear:button-clear'
   ],
 
-  kindClass: Ember.computed('kind', function() {
-    let kind = this.get('kind');
-    if (kind) {
-      return `button-${this.get('kind')}`;
+  kindClass: Ember.computed('kind', {
```
Comment (ebryn) [inline addon/components/ion-button.js]: revert back to older CP syntax for compatibility

File: addon/components/ion-item.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import IonList from './ion-list';
+
+Ember.LinkView.reopen({
+  attributeBindings: ["style"]
+});
+
+export default Ember.Component.extend({
+  layout: layout,
+  tagName: 'ion-item',
+  classNames: ['item', 'item-remove-animate', 'item-avatar', 'item-icon-right', 'item-complex', 'item-right-editable'],
```
Comment (ebryn) [inline addon/components/ion-item.js]: many of these are supposed to be conditional

File: addon/components/ion-item.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  }),
+
+  registerOptionButton(button) {
+    this.get('optionButtons').pushObject(button);
+  },
+
+  didInsertElement() {
+    let {width} = this.element.getBoundingClientRect();
+    let hammer = new Hammer(this.element);
+    let slidingLink = this.element.getElementsByClassName("item-content")[0];
```
Comment (ebryn) [inline addon/components/ion-item.js]: use `querySelector`

File: addon/components/ion-item.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      if (event.type === 'pan') {
+        this.pan(event);
+      } else if (event.type === 'panend') {
+        this.panEnd(event);
+      }
+    });
+
+    slidingLink.addEventListener("transitionend", () => {
+      if (!this.open) {
+        this.element.getElementsByClassName("item-options")[0].classList.add("invisible");
```
Comment (ebryn) [inline addon/components/ion-item.js]: querySelector

File: addon/components/ion-item.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+  init() {
+    this._super(...arguments);
+    let ionList = this.nearestOfType(IonList);
+    ionList.registerItem(this);
+  },
+
+  translateX: 0,
+  originalWidth: 0,
+  translateSpeed: 0,
```
Comment (JFickel) [inline addon/components/ion-item.js]: possibly remove

File: addon/components/ion-item.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  registerOptionButton(button) {
+    this.get('optionButtons').pushObject(button);
+  },
+
+  didInsertElement() {
+    let {width} = this.element.getBoundingClientRect();
+    let hammer = new Hammer(this.element);
+    let slidingLink = this.element.querySelector(".item-content");
+
+    this.originalWidth = width;
```
Comment (JFickel) [inline addon/components/ion-item.js]: rename to width

File: addon/components/ion-item.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  init() {
+    this._super(...arguments);
+    let ionList = this.nearestOfType(IonList);
+    ionList.registerItem(this);
+  },
+
+  translateX: 0,
+  originalWidth: 0,
+  translateSpeed: 0,
+  open: null,
```
Comment (JFickel) [inline addon/components/ion-item.js]: set to false and underscore internals

File: addon/components/ion-item.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  optionVisibility: 'invisible',
```
Comment (JFickel) [inline addon/components/ion-item.js]: optionVisibilityClass*

File: addon/components/ion-item.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+  transformOutput: Ember.computed('translateX', 'translateSpeed', function() {
+    return Ember.String.htmlSafe(`transform: translate(${this.get('translateX')}px, 0px);
```
Comment (JFickel) [inline addon/components/ion-item.js]: let escape = Ember.Handlebars.Utils.escapeExpression;

File: addon/components/ion-item.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  },
+
+  didInsertElement() {
+    let {width} = this.element.getBoundingClientRect();
+    let hammer = new Hammer(this.element);
+    let slidingLink = this.element.querySelector(".item-content");
+
+    this.originalWidth = width;
+
+    hammer.on('pan panend', event => {
```
Comment (JFickel) [inline addon/components/ion-item.js:63]: remind Erik to teach me about custom event delegation with the event dispatcher

File: addon/components/ion-item.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+  panEnd(event) {
+    Ember.run(() => {
+      this.finishSlidingItem(event);
+    });
+  },
+
+  slideItem(event) {
+    let offset;
+    if (!this.originX) {
```
Comment (JFickel) [inline addon/components/ion-item.js]: declare originX on the class. maybe rename to previousX

File: addon/components/ion-item.js [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      this.set('optionVisibility', '');
+      offset = this.originX = event.deltaX;
+    } else {
+      offset = event.deltaX - this.originX;
+      this.originX = event.deltaX;
+    }
+    this.moveAtIndex(offset);
```
Comment (JFickel) [inline addon/components/ion-item.js]: rename to move

File: addon/components/ion-item.js [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+  },
+
+  moveAtIndex(offset) {
+    let withinRightMax = (this.translateX + offset) <= 0;
+    let withinLeftMax = (this.translateX + offset) >= (this.originalWidth * -0.70);
```
Comment (JFickel) [inline addon/components/ion-item.js]: don't use magic numbers. use const

File: addon/components/ion-item.js [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    if (withinRightMax && withinLeftMax) {
+      this.set('translateX', this.translateX + offset);
+    }
+  },
+
+  finishSlidingItem() {
+    let optionsWidth = this.element.querySelector(".item-options").offsetWidth;
+    this.set('translateSpeed', 0.4);
```
Comment (JFickel) [inline addon/components/ion-item.js]: const as well

File: addon/components/ion-item.js [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+  moveAtIndex(offset) {
+    let withinRightMax = (this.translateX + offset) <= 0;
+    let withinLeftMax = (this.translateX + offset) >= (this.originalWidth * -0.70);
+    if (withinRightMax && withinLeftMax) {
+      this.set('translateX', this.translateX + offset);
+    }
+  },
+
+  finishSlidingItem() {
+    let optionsWidth = this.element.querySelector(".item-options").offsetWidth;
```
Comment (JFickel) [inline addon/components/ion-item.js:114]: what's the fastest way of getting the width? offset/boundingclient?

File: addon/components/ion-item.js [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+  open: null,
+  optionVisibility: 'invisible',
+
+  transformOutput: Ember.computed('translateX', 'translateSpeed', function() {
+    return Ember.String.htmlSafe(`transform: translate(${this.get('translateX')}px, 0px);
+            transition: ${this.get('translateSpeed')}s transform`);
+  }),
+
+  optionButtons: Ember.computed(function() {
+    return Ember.A([]);
```
Comment (JFickel) [inline addon/components/ion-item.js]: set in init with underscore

File: addon/components/ion-option-button.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import Ember from 'ember';
+import layout from '../templates/components/ion-option-button';
+import IonItem from './ion-item';
+
+export default Ember.Component.extend({
+  layout: layout,
+  classNames: ['button'],
```
Comment (JFickel) [inline addon/components/ion-option-button.js]: delete

File: addon/components/ion-option-button.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+  tagName: '',
+  kindClass: Ember.computed('kind', function() {
+    let kind = this.get('kind');
+    if (kind) {
+      return `button button-${kind}`;
```
Comment (JFickel) [inline addon/components/ion-option-button.js]: conditionally append the kind class. should always have button

File: addon/components/ion-radio.js [hunk 19]
```diff
@@ -0,0 +1,8 @@
+import Ember from 'ember';
+import layout from '../templates/components/ion-radio';
+
+export default Ember.Component.extend({
+  layout: layout,
+  tagName: 'label',
+  classNames: ['radio']
```
Comment (JFickel) [inline addon/components/ion-radio.js]: implement or remove

File: addon/components/ion-slide-box.js [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    currentSlide.set('translateX', currentSlide.get('translateX') + offset);
+
+    if (restoreNext && nextSlide && offset > 0) {
+      nextSlide.set('translateX', nextSlide.get('translateX') + restoreNext);
+    } else if (nextSlide) {
+      nextSlide.set('translateX', nextSlide.get('translateX') + offset);
+    }
+  },
+
+  setupSlides() {
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: rename to stackSlides

File: addon/components/ion-slide-box.js [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+      restorePrevious = -deltaX;
+    } else {
+      offset = -deltaX;
+    }
+
+    this.moveAtIndex(offset, currentSlideIdx, restoreNext, restorePrevious);
+
+  },
+
+  translateSlides(event) {
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: pass deltaX instead of event

File: addon/components/ion-slide-box.js [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+  pan(event) {
+    Ember.run(() => {
+      this.translateSlides(event);
+    });
+  },
+
+  panEnd(event) {
+    Ember.run(() => {
+      this.originX = null;
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: define in class

File: addon/components/ion-slide-box.js [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+      this.finishTranslation(event.deltaX);
+    });
+  },
+
+  finishTranslation(deltaX) {
+    let threshold = this.originalWidth/2;
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: try to copy logic closely

File: addon/components/ion-slide-box.js [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    let offset, restoreNext, restorePrevious;
+    let currentSlideIdx = this.currentSlideIdx;
+
+    if (deltaX > threshold && currentSlideIdx > 0) {
+      // move index to previous slide
+      this.decrementProperty('currentSlideIdx');
+      offset = this.originalWidth - deltaX;
+      // these restore variables return the slides to the "deck"
+      restoreNext = -deltaX;
+    } else if (deltaX < -threshold && currentSlideIdx < (this._slides.length - 1)) {
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: move `this._slides.length - 1`into var

File: addon/components/ion-slide-box.js [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+      offset = this.originX = event.deltaX;
+    } else {
+      offset = event.deltaX - this.originX;
+      this.originX = event.deltaX;
+    }
+
+    this.moveAtIndex(offset, this.currentSlideIdx);
+  },
+
+  moveAtIndex(offset, currentIndex, restoreNext, restorePrevious) {
```
Comment (JFickel) [inline addon/components/ion-slide-box.js]: name above to compute translation and this method to  _translate

File: addon/components/ion-slide.js [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+import layout from '../templates/components/ion-slide';
+import IonSlideBox from './ion-slide-box';
+
+export default Ember.Component.extend({
+  layout: layout,
+  tagName: 'ion-slide',
+  classNames: ['slider-slide'],
+  attributeBindings: ['style'],
+  style: Ember.computed('width', 'height', 'left', 'translateX', function() {
+    return Ember.String.htmlSafe(
```
Comment (JFickel) [inline addon/components/ion-slide.js:19]: sanitize

File: addon/components/ion-toggle.js [hunk 27]
```diff
@@ -3,8 +3,8 @@ import layout from '../templates/components/ion-toggle';
 
 export default Ember.Component.extend({
   layout: layout,
-  tagName: 'label',
-  classNames: ['toggle'],
+  tagName: 'div',
+  classNames: ['item', 'item-toggle', 'toggle-large'],
```
Comment (JFickel) [inline addon/components/ion-toggle.js:7]: look into kind

**Pull request merged**